### PR TITLE
Create template based identity provider with authentication configurations

### PR DESCRIPTION
**Pull request by tharindu-bandara**
## Purpose
- This PR addresses the issue: https://github.com/wso2/identity-apps/issues/354.

## Goals
- Implement a basic flow to add an identity provider(authentication only) via a template-based Wizard.
- Support any available authenticator in the server(SAML, OIDC, Google, Facebook etc.)

## Approach
- Use hard-coded templates in the `Add Identity Provider` page.
Screenshot from 2020-03-28 17-10-04
- The wizard has three steps; General Details, Authenticator, and Summary.
Screenshot from 2020-03-28 17-11-07
- Federated authenticator related settings are configured in the authenticator step. Server metadata APIs are used to dynamically generate configuration form.
Screenshot from 2020-03-29 02-49-04

- In the end, a summary is shown prior to submitting the create request.
Screenshot from 2020-03-28 17-16-37
- Upon clicking the finish button, the IDP is created and the user is redirected to its edit page(already existing implementation).
Screenshot from 2020-03-28 17-23-29

### CODE REVIEW

File: apps/admin-portal/src/api/identity-provider.ts [hunk 1]
```diff
@@ -34,6 +34,34 @@ import {
  */
 const httpClient = AxiosHttpClient.getInstance();
 
+/**
+ * Creates a new identity provider.
```
Comment (jeradrutnam) [inline apps/admin-portal/src/api/identity-provider.ts]:
```suggestion
 * Creates Identity Provider.
```

File: apps/admin-portal/src/api/identity-provider.ts [hunk 2]
```diff
@@ -262,3 +290,62 @@ export const getFederatedAuthenticatorMeta = (id: string): Promise<any> => {
...
+        url: ServiceResourcesEndpoint.identityProviders + "/meta/federated-authenticators/" + authenticatorId
+    };
+
+    return httpClient.request(requestConfig)
+        .then((response) => {
+            if (response.status !== 200) {
+                return Promise.reject(new Error("Failed to get federated authenticator metadata for: "
+                    + authenticatorId));
+            }
+            return Promise.resolve(response.data as FederatedAuthenticatorMetaInterface);
```
Comment (jeradrutnam) [inline apps/admin-portal/src/api/identity-provider.ts]:
```suggestion

            return Promise.resolve(response.data as FederatedAuthenticatorMetaInterface);
```

File: apps/admin-portal/src/components/identityProviders/forms/authenticators/common-authenticator-form.tsx [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+ */
+enum CommonAuthenticatorConstants {
+    BOOLEAN = "BOOLEAN"
+}
+
+/**
+ * Google authenticator configurations form.
+ *
+ * @param {AuthenticatorFormPropsInterface} props
+ * @return { ReactElement }
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/identityProviders/forms/authenticators/common-authenticator-form.tsx]:
```suggestion
 * @return { FunctionComponent }
```

File: apps/admin-portal/src/components/identityProviders/identity-provider-list.tsx [hunk 4]
```diff
@@ -24,6 +24,7 @@ import { useDispatch } from "react-redux";
 import { history } from "../../helpers";
 import { deleteIdentityProvider } from "../../api";
 import { IdentityProviderListResponseInterface } from "../../models";
+import {GlobalConfig} from "../../configs";
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/identityProviders/identity-provider-list.tsx]:
```suggestion
import { GlobalConfig } from "../../configs";
```

File: apps/admin-portal/src/components/identityProviders/wizard/steps/wizard-summary.tsx [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { AppAvatar, Heading } from "@wso2is/react-components";
+import React, { FunctionComponent, ReactElement, useEffect } from "react";
+import { Divider, Grid } from "semantic-ui-react";
+import {FederatedAuthenticatorMetaInterface, IdentityProviderInterface} from "../../../../models";
```
Comment (jeradrutnam) [inline apps/admin-portal/src/components/identityProviders/wizard/steps/wizard-summary.tsx]:
```suggestion
import { FederatedAuthenticatorMetaInterface, IdentityProviderInterface } from "../../../../models";
```

File: apps/admin-portal/src/models/identity-provider.ts [hunk 6]
```diff
@@ -16,6 +16,8 @@
  * under the License.
  */
 
+import {AuthProtocolMetaListItemInterface} from "./application-inbound";
```
Comment (jeradrutnam) [inline apps/admin-portal/src/models/identity-provider.ts]:
```suggestion
import { AuthProtocolMetaListItemInterface } from "./application-inbound";
```

File: apps/admin-portal/src/pages/identity-provider-template.tsx [hunk 7]
```diff
@@ -16,9 +16,9 @@
  * under the License.
  */
 
-import React, { FunctionComponent, SyntheticEvent, useState } from "react";
+import React, { FunctionComponent, ReactElement, SyntheticEvent, useState } from "react";
 import { QuickStartIdentityProviderTemplates } from "../components";
-import { IdPIcons, IdPCapabilityIcons } from "../configs";
+import {IdPCapabilityIcons, IdPIcons, InboundProtocolLogos} from "../configs";
```
Comment (jeradrutnam) [inline apps/admin-portal/src/pages/identity-provider-template.tsx]:
```suggestion
import { IdPCapabilityIcons, IdPIcons, InboundProtocolLogos } from "../configs";
```

File: apps/admin-portal/src/store/actions/types/identityProvider.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+ * under the License.
+ */
+
+/**
+ * Enum for identity provider action types.
+ *
+ * @readonly
+ * @enum {string}
+ */
+import {FederatedAuthenticatorListItemInterface, FederatedAuthenticatorMetaInterface} from "../../../models";
```
Comment (jeradrutnam) [inline apps/admin-portal/src/store/actions/types/identityProvider.ts]:
```suggestion
import { FederatedAuthenticatorListItemInterface, FederatedAuthenticatorMetaInterface } from "../../../models";
```

File: apps/admin-portal/src/store/combine-reducers.ts [hunk 9]
```diff
@@ -16,16 +16,18 @@
  * under the License.
  */
 
-import { combineReducers } from "redux";
-import { applicationReducer, authenticateReducer, globalReducer } from "./reducers";
-import { LoadersReducer } from "./reducers/loaders";
+import {combineReducers} from "redux";
+import {applicationReducer, authenticateReducer, globalReducer} from "./reducers";
+import {LoadersReducer} from "./reducers/loaders";
+import {identityProviderReducer} from "./reducers/identityProvider";
```
Comment (jeradrutnam) [inline apps/admin-portal/src/store/combine-reducers.ts]:
```suggestion
import {combineReducers} from "redux";
import { applicationReducer, authenticateReducer, globalReducer } from "./reducers";
import { LoadersReducer } from "./reducers/loaders";
import { identityProviderReducer } from "./reducers/identityProvider";
```

File: apps/admin-portal/src/store/reducers/identityProvider.ts [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import {IdentityProviderReducerStateInterface} from "../../models";
+import {IdentityProviderActions, IdentityProviderActionTypes} from "../actions/types/identityProvider";
```
Comment (jeradrutnam) [inline apps/admin-portal/src/store/reducers/identityProvider.ts]:
```suggestion
import { IdentityProviderReducerStateInterface } from "../../models";
import { IdentityProviderActions, IdentityProviderActionTypes } from "../actions/types/identityProvider";
```

Comment (jeradrutnam) [inline apps/admin-portal/src/store/reducers/identityProvider.ts]: Shall we make the file lowercase ?

File: apps/admin-portal/src/utils/identity-provider-management-utils.ts [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    AuthProtocolMetaListItemInterface,
+    FederatedAuthenticatorListItemInterface,
+    FederatedAuthenticatorMetaInterface
+} from "../models";
+import {getFederatedAuthenticatorsList} from "../api";
+import _ from "lodash";
+import {store} from "../store";
+import {addAlert} from "@wso2is/core/store";
+import {AlertLevels} from "@wso2is/core/models";
+import {setAvailableAuthenticatorsMeta} from "../store/actions/IdentityProvider";
```
Comment (jeradrutnam) [inline apps/admin-portal/src/utils/identity-provider-management-utils.ts]:
```suggestion
import { getFederatedAuthenticatorsList } from "../api";
import _ from "lodash";
import { store } from "../store";
import { addAlert } from "@wso2is/core/store";
import { AlertLevels } from "@wso2is/core/models";
import { setAvailableAuthenticatorsMeta } from "../store/actions/IdentityProvider";
```

File: apps/admin-portal/src/utils/identity-provider-management-utils.ts [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+    /**
+     * Private constructor to avoid object instantiation from outside
+     * the class.
+     *
+     * @hideconstructor
+     */
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    private constructor() {
+    }
```
Comment (jeradrutnam) [inline apps/admin-portal/src/utils/identity-provider-management-utils.ts]:
```suggestion
    private constructor() { }
```

**Comment by jeradrutnam** [comment]
Shall we make these file to lowercase ?

apps/admin-portal/src/store/actions/IdentityProvider.ts
apps/admin-portal/src/store/actions/types/identityProvider.ts
apps/admin-portal/src/store/reducers/identityProvider.ts

**Comment by tharindu-bandara** [comment]
> Shall we make these file to lowercase ?
> 
> apps/admin-portal/src/store/actions/IdentityProvider.ts
> apps/admin-portal/src/store/actions/types/identityProvider.ts
> apps/admin-portal/src/store/reducers/identityProvider.ts

Fixed with https://github.com/wso2/identity-apps/pull/632/commits/e0c193e52bcf9374036f406699147fbc98e8bd77.

File: apps/admin-portal/src/store/reducers/identityProvider.ts (hunk 10 above)
Comment (tharindu-bandara) [reply apps/admin-portal/src/store/reducers/identityProvider.ts]: Fixed with https://github.com/wso2/identity-apps/pull/632/commits/e0c193e52bcf9374036f406699147fbc98e8bd77.

**Pull request merged**